_posts/blog/2024-04-08-does-a-compactor-return-memory-to-OS.md

```diff
@@ -0,0 +1,232 @@
+   try { 
+       client.tableOperations().delete(tableName); 
+   } catch (TableNotFoundException e) { 
+       // ignore 
+   } 
+   
+   System.out.println("Creating table " + tableName); 
+   client.tableOperations().create(tableName); 
+   
+   // This is done to avoid system compactions, we want to initiate the compactions manually 
+   client.tableOperations().setProperty(tableName, Property.TABLE_MAJC_RATIO.getKey(), "1000"); 
+   // Configure for external compaction 
+   client.instanceOperations().setProperty("tserver.compaction.major.service.cs1.planner","org.apache.accumulo.core.spi.compaction.DefaultCompactionPlanner"); 
+   client.instanceOperations().setProperty("tserver.compaction.major.service.cs1.planner.opts.executors","[{\"name\":\"large\",\"type\":\"external\",\"queue\":\"q1\"}]"); 
+   
+   client.tableOperations().setProperty(tableName, "table.compaction.dispatcher", "org.apache.accumulo.core.spi.compaction.SimpleCompactionDispatcher"); 
+   client.tableOperations().setProperty(tableName, "table.compaction.dispatcher.opts.service", "cs1"); 
+   
+   int numFiles = 20; 
+   
+   try (var writer = client.createBatchWriter(tableName)) { 
+       for (int i = 0; i < numFiles; i++) { 
+           Mutation mut = new Mutation("r" + i); 
+           mut.at().family("cf").qualifier("cq").put(data); 
+           writer.addMutation(mut); 
+           writer.flush();   
+   
+           System.out.println("Writing " + dataSize + " bytes to a single value"); 
+           client.tableOperations().flush(tableName, null, null, true); 
+       } 
+   }   
+   
+   System.out.println("Compacting table"); 
+   client.tableOperations().compact(tableName, new CompactionConfig().setWait(true)); 
+   System.out.println("Finished table compaction");
+} 
+
+ingestAndCompact(); 
+// Optionally sleep and ingestAndCompact() again, or just execute the script again.
+```
+
+### OS Memory Data Collection Script
+***referred to as mem_usage_script.sh in the test setup section***
+```bash
+#!/bin/bash 
+echo "usage: set PID in script to the compactor PID then run." 
+PID=xxxxx     # NOTE: Must set PID 
+echo "Tracking PID: $PID" 
+DURATION=3600 # for 1 hour 
+INTERVAL=5    # every 5 seconds 
+rm output_mem_usage.log 
+
+while [ $DURATION -gt 0 ]; do 
+    ps -o %mem,rss -p $PID | tail -n +2 >> output_mem_usage.log 
+    sleep $INTERVAL 
+    DURATION=$((DURATION - INTERVAL)) 
+done
+```
+
+After compactions have completed plot the data using gnuplot:
+
+```bash
+gnuplot
+set title "Resident Set Size (RSS) Memory usage" 
+set xlabel "Time"
+set ylabel "Mem usage in kilobytes"
+plot "output_mem_usage.log" using ($0*5):2 with lines title 'Mem usage'
+```
+
+## Data
+
+Important Notes:
+- ZGC and G1PeriodicGCInterval are not available with Java 11 so couldn't be tested for
+- The only tests which used manual GC are those labeled as such
+- G1 GC is the default GC in Java 11, 17, and 21 (doesn't need to be specified in java args)
+
```

Comment:
I wonder if there should be a table here that summarizes the JVM version, GC used, and JVM options.